src/libraries/System.Net.Quic/tests/FunctionalTests/MsQuicTests.cs

```diff
@@ -388,58 +388,133 @@ public async Task ConnectWithClientCertificate(bool sendCerttificate)
             (QuicConnection clientConnection, QuicConnection serverConnection) = await CreateConnectedQuicConnection(null, listenerOptions);
 
-            // No stream opened yet, should return immediately.
-            Assert.True(clientConnection.WaitForAvailableBidirectionalStreamsAsync().IsCompletedSuccessfully);
+            CancellationTokenSource cts = new CancellationTokenSource();
 
-            // Open one stream, should wait till it closes.
-            QuicStream stream = clientConnection.OpenBidirectionalStream();
-            ValueTask waitTask = clientConnection.WaitForAvailableBidirectionalStreamsAsync();
+            // Open one stream, second call should block
+            QuicStream stream = await OpenStreamAsync(clientConnection);
+            ValueTask<QuicStream> waitTask = OpenStreamAsync(clientConnection, cts.Token);
             Assert.False(waitTask.IsCompleted);
-            Assert.Throws<QuicException>(() => clientConnection.OpenBidirectionalStream());
+
+            cts.Cancel();
+
+            // awaiting the task should throw
+            var ex = await Assert.ThrowsAsync<OperationCanceledException>(() => waitTask.AsTask().WaitAsync(TimeSpan.FromSeconds(3)));
+            Assert.Equal(cts.Token, ex.CancellationToken);
 
             // Close the streams, the waitTask should finish as a result.
             stream.Dispose();
             QuicStream newStream = await serverConnection.AcceptStreamAsync();
             newStream.Dispose();
-            await waitTask.AsTask().WaitAsync(TimeSpan.FromSeconds(10));
+
+            // next call should work as intended
+            newStream = await OpenStreamAsync(clientConnection).AsTask().WaitAsync(TimeSpan.FromSeconds(10));
+            newStream.Dispose();
+
+            clientConnection.Dispose();
+            serverConnection.Dispose();
+        }
+
+        [Theory]
+        [InlineData(false)]
+        [InlineData(true)]
+        public async Task OpenStreamAsync_PreCanceled_Throws_OperationCanceledException(bool unidirectional)
+        {
+            ValueTask<QuicStream> OpenStreamAsync(QuicConnection connection, CancellationToken token = default) => unidirectional
+                ? connection.OpenUnidirectionalStreamAsync(token)
+                : connection.OpenBidirectionalStreamAsync(token);
+
+            (QuicConnection clientConnection, QuicConnection serverConnection) = await CreateConnectedQuicConnection(null, CreateQuicListenerOptions());
+
+            CancellationTokenSource cts = new CancellationTokenSource();
+            cts.Cancel();
+
+            var ex = await Assert.ThrowsAsync<OperationCanceledException>(() => OpenStreamAsync(clientConnection, cts.Token).AsTask().WaitAsync(TimeSpan.FromSeconds(3)));
+            Assert.Equal(cts.Token, ex.CancellationToken);
+
             clientConnection.Dispose();
             serverConnection.Dispose();
         }
 
+        [Theory]
+        [InlineData(false, false)]
+        [InlineData(true, true)]
```

Comment:
should we have a "full matrix" here (4 InlineData's)?

Reply:
We can, but it does not add any real value IMO